### ISSUE TITLE
Rule proposal: Require that generic is specified

### ISSUE THREAD

**Issue by jsphstls** [typescript-eslint/typescript-eslint#4391]
### Description

Generics are beneficial to use and not using them can introduce confusion by potentially misusing an API, since the presence of a generic indicates its intended usage.

There is no official way to enforce that generic type values are supplied in TypeScript despite the [popularity](https://stackoverflow.com/questions/51173191/typescript-require-generic-parameter-to-be-provided) of the [request](https://github.com/Microsoft/TypeScript/issues/14829).

As a real-world example, the https://github.com/downshift-js/downshift React component allows for a generic to be specified for the component instance. Specifying that generic allows other types to be inferred, thereby simplifying the code.

### Fail

```ts

const someFunction = <SomeGeneric>(someArg: SomeGeneric) => null

someFunction(someObject)
```

### Pass

```ts
const someFunction = <SomeGeneric>(someArg: SomeGeneric) => null

someFunction<SomeInterface>(someObject)

// eslint-disable-next-line @typescript-eslint - I choose not to supply the generic.
someFunction(someObject)
```

**Comment by bradzacher** [comment]
Thanks for the request!

One of the criteria we look at when adding a lint rule to this project is the "rate of false-positives". Ideally this would be 0 - a rule never false positives, but sometimes we allow *some* false-positives because they only occur in very rare cases.

For this request the false positive rate would likely be close to 100% - in almost all cases you would want to use TS's inferred generic types instead of supplying your own. This would lead to codebases either becoming inundated with disable comments, or in some cases it could lead to users supplying sub-optimal types instead of importing types from modules.

Another criteria is the "general applicability" of a rule - do we think that a substantial percentage of our users find value from such a rule. IMO based on the above - most users would not use this rule as it would make their codebase too cumbersome to work in.

For these reasons - we won't be accepting this rule in this project.
Our tooling is available for anyone to use to build rules for their own eslint plugins, so feel free to build and publish if you will find value in it!
https://typescript-eslint.io/docs/development/custom-rules
